### PR TITLE
feat(wallet): Serliaze and deserialize numbers to strings

### LINKED ISSUE #1760: [Wallet] Send wits as String instead of Number
The wallet is sending values greater than the `Number.MAX_SAFE_INTEGER` which produces a wrong behavior in Sheikah. To avoid transform all the calls related to the wallet and double this code across multiple clients, it could use a macro to send and receive Strings in the following calls:

- [ ] create_data_request
- [ ] create_vtt
- [ ] get_balance
- [ ] get_transactions
- [ ] send_transaction

and events:
- [ ] Movement

### PR DESCRIPTION
**Pull request by clbartoli**
This PR de/serializes numbers in components sent to Sheikah through wallet methods:
- get_balance
- get_transaction
- get_addresses
- create_vtt
- create_data_request

This way accept numbers as strings and sends numbers as strings
The PR can be tested using Sheikhas's PR https://github.com/witnet/sheikah/pull/1577

Closes #1760

### CODE REVIEW

File: wallet/src/actors/app/handlers/create_vtt.rs [hunk 1]
```diff
@@ -128,6 +145,255 @@ pub fn validate_output_addresses(
...
+    Commit(CommitTransaction),
+    Reveal(RevealTransaction),
+    Tally(TallyTransaction),
+    Mint(MintTransaction),
+}
+
+impl From<Transaction> for TransactionHelper {
+    fn from(x: Transaction) -> Self {
+        match x {
+            Transaction::ValueTransfer(vttransaction) => {
```
Comment (girazoki) [inline wallet/src/actors/app/handlers/create_vtt.rs:202]: Isnt this repeated code? If so, can we just move it to a single file?

File: wallet/src/model.rs [hunk 2]
```diff
@@ -1,8 +1,11 @@
 //! Types that are serializable and can be returned as a response.
 use std::{collections::HashMap, fmt};
 
+use std::fmt::Display;
+use std::str::FromStr;
```
Comment (girazoki) [inline wallet/src/model.rs:5]: FMT

File: wallet/src/model.rs [hunk 3]
```diff
@@ -105,12 +120,17 @@ where
 pub struct BalanceMovement {
     /// Database key for storing `BalanceMovement` objects
     #[serde(skip)]
+    // #[serde(serialize_with = "u32_to_string", deserialize_with = "number_from_string")]
```
Comment (lrubiorod) [inline wallet/src/model.rs:123]: why this comment?

File: wallet/src/types.rs [hunk 4]
```diff
@@ -275,3 +297,161 @@ pub struct SuperBlockNotification {
...
+    U: From<T>,
+    U: Serialize,
+{
+    let x = U::from(val.clone());
+    x.serialize(serializer)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::app::CreateDataReqRequest;
```
Comment (lrubiorod) [inline wallet/src/types.rs:445]: fmt

File: wallet/src/actors/app/handlers/create_vtt.rs (hunk 1 above)
Comment (tmpolaczyk) [reply wallet/src/actors/app/handlers/create_vtt.rs:210]: Something like `wallet/src/serialization.rs`

File: wallet/src/model.rs (hunk 3 above)
Comment (tmpolaczyk) [reply wallet/src/model.rs:123]: Not needed

```suggestion
```

File: wallet/src/model.rs [hunk 5]
```diff
@@ -144,8 +164,13 @@ pub struct Transaction {
     /// Transaction hash (used as identifier)
     pub hash: String,
     /// Reward to miner for including transaction in the block
+    #[serde(
+        serialize_with = "u64_to_string",
+        deserialize_with = "number_from_string"
+    )]
     pub miner_fee: u64,
     /// Date when transaction was included a block (same as block date)
+    // #[serde(serialize_with = "u64_to_string", deserialize_with = "number_from_string")]
```
Comment (tmpolaczyk) [inline wallet/src/model.rs:173]: So timestamps are numbers or strings?

File: wallet/src/model.rs [hunk 6]
```diff
@@ -191,13 +216,22 @@ pub struct MintData {
...
+        serialize_with = "u64_to_string",
+        deserialize_with = "number_from_string"
+    )]
     pub value: u64,
 }
 
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
 pub struct Output {
     pub address: String,
+    //#[serde(serialize_with = "u64_to_string", deserialize_with = "number_from_string")]
```
Comment (lrubiorod) [inline wallet/src/model.rs:229]: remove comment?

File: wallet/src/model.rs [hunk 7]
```diff
@@ -217,12 +251,17 @@ pub struct Reveal {
...
+        serialize_with = "u32_to_string",
+        deserialize_with = "number_from_string"
+    )]
     pub total: u32,
 }
 
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, Hash)]
 pub struct OutPtr {
     pub txn_hash: Vec<u8>,
+    //#[serde(serialize_with = "u32_to_string", deserialize_with = "number_from_string")]
```
Comment (lrubiorod) [inline wallet/src/model.rs:264]: remove comment?

File: wallet/src/types.rs [hunk 8]
```diff
@@ -275,3 +298,512 @@ pub struct SuperBlockNotification {
...
+//         DataRequestOutput {
+//             data_request: x.data_request,
+//             witness_reward: x.witness_reward,
+//             witnesses: x.witnesses,
+//             commit_and_reveal_fee: x.commit_and_reveal_fee,
+//             min_consensus_percentage: x.min_consensus_percentage,
+//             collateral: x.collateral,
+//         }
+//     }
+// }
```
Comment (tmpolaczyk) [inline wallet/src/types.rs:695]: What about this commented code?

File: wallet/src/types.rs [hunk 9]
```diff
@@ -275,3 +298,512 @@ pub struct SuperBlockNotification {
...
+    }
+
+    #[test]
+    fn test_desarialize_create_dr() {
+        let _e1: CreateDataReqRequest = serde_json::from_str(r#"{"session_id":"a273d76ab9870976ab82a5dddced02fcd3ab5ff4fac51a1cf9437c762f317101","wallet_id":"87575c9031c01cf84dffc33fe2d28474d620dacd673f06990dc0318079ddfde7","fee":"1","request":{"data_request":{"timeLock":0,"time_lock":0,"retrieve":[{"contentType":"JSON API","kind":"HTTP-GET","script":[128],"url":"https://blockchain.info/q/latesthash"},{"contentType":"JSON API","kind":"HTTP-GET","script":[130,24,119,130,24,103,100,104,97,115,104],"url":"https://api-r.bitcoinchain.com/v1/status"},{"contentType":"JSON API","kind":"HTTP-GET","script":[131,24,119,130,24,102,100,100,97,116,97,130,24,103,111,98,101,115,116,95,98,108,111,99,107,95,104,97,115,104],"url":"https://api.blockchair.com/bitcoin/stats"}],"aggregate":{"filters":[],"reducer":2},"tally":{"filters":[{"op":8,"args":[]}],"reducer":2}},"collateral":"1000000000","witness_reward":"1","witnesses":"3","commit_and_reveal_fee":"1","min_consensus_percentage":"51"}}"#).unwrap();
+    }
```
Comment (tmpolaczyk) [inline wallet/src/types.rs:808]: Add more tests for the other methods?

File: wallet/src/types.rs [hunk 10]
```diff
@@ -2,11 +2,16 @@ use std::convert::TryFrom;
...
+use core::fmt::Display;
+pub use std::str::FromStr;
+
 use failure::Fail;
 pub use jsonrpc_core::{Params as RpcParams, Value as RpcValue};
 pub use jsonrpc_pubsub::{Sink, SinkResult, Subscriber, SubscriptionId};
-use serde::{Deserialize, Serialize};
+use serde::{Deserialize, Deserializer, Serialize};
 pub use serde_json::Value as Json;
+//use crate::app::CreateDataReqRequest;
```
Comment (tmpolaczyk) [inline wallet/src/types.rs:13]:
```suggestion
```

File: wallet/src/actors/app/handlers/create_data_req.rs [hunk 11]
```diff
@@ -3,32 +3,52 @@ use serde::{Deserialize, Serialize};
 
 use crate::{
     actors::app,
-    types::{self, Hashable as _, ProtobufConvert as _},
+    types::{
+        self, from_generic_type, into_generic_type, number_from_string, u64_to_string,
+        DataRequestOutputHelper, Hashable as _, ProtobufConvert as _, TransactionHelper,
+    },
+};
+pub use witnet_data_structures::{
```
Comment (tmpolaczyk) [inline wallet/src/actors/app/handlers/create_data_req.rs:11]: `pub use` means "import and re-export", we only need to import:

```suggestion
use witnet_data_structures::{
```

File: wallet/src/model.rs [hunk 12]
```diff
@@ -5,7 +5,10 @@ use failure::_core::fmt::Formatter;
 use serde::{Deserialize, Serialize};
 
 use crate::repository::keys::Key;
-use crate::{account, types};
+use crate::{
```
Comment (lrubiorod) [inline wallet/src/model.rs:8]: fmt

File: wallet/src/types.rs [hunk 13]
```diff
@@ -275,3 +297,437 @@ pub struct SuperBlockNotification {
...
+            value: x.value,
+            time_lock: x.time_lock,
+        }
+    }
+}
+
+/// Value transfer output transaction data structure
+#[derive(Debug, Eq, PartialEq, Clone, Serialize, Deserialize, Hash, Default)]
+pub struct VttOutputParamsHelper {
+    pub address: String,
```
Comment (tmpolaczyk) [inline wallet/src/types.rs:570]: I think this should use the `PublicKeyHash` type, this way invalid addresses are found at deserialization time:

```suggestion
    pub address: PublicKeyHash,
```

**Review by lrubiorod — APPROVED** [review]

File: wallet/src/types.rs [hunk 14]
```diff
@@ -151,14 +155,32 @@ pub struct CreateWalletData<'a> {
     pub master_key: Option<ExtendedSK>,
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize)]
```
Comment (tmpolaczyk) [inline wallet/src/types.rs:158]: What happens if we revert the changes to this struct?

```suggestion
```

**Review by tmpolaczyk — APPROVED** [review]

**Review by aesedepece — APPROVED** [review]
